android/src/main/kotlin/media/bcc/bccm_player/Downloader.kt

```diff
@@ -0,0 +1,246 @@
+
+fun createDownloadManager(context: Context): DownloadManager {
+    val databaseProvider = StandaloneDatabaseProvider(context)
+    val cacheDir = File(context.filesDir, DOWNLOAD_FOLDER_NAME)
+    if (cache == null) {
+        cache = SimpleCache(cacheDir, NoOpCacheEvictor(), databaseProvider)
+    }
+    val dataSourceFactory = DefaultHttpDataSource.Factory()
+    val downloadExecutor = Executor(Runnable::run)
+    return DownloadManager(context, databaseProvider, cache!!, dataSourceFactory, downloadExecutor)
+}
+
+@Parcelize
+data class DownloadInfo(
+    val title: String,
+    val audioTrackIds: List<String>,
+    val videoTrackIds: List<String>,
+    val additionalData: Map<String, String>
+) : Parcelable
+
+object ParcelMarshall {
+    fun <T> pack(input: T): ByteArray where T : Parcelable {
+        val dataParcel = Parcel.obtain()
+        input.writeToParcel(dataParcel, 0)
+        val data = dataParcel.marshall()
+        dataParcel.recycle()
+        return data
+    }
+
+    fun <T> unpack(byteArray: ByteArray, creator: Parcelable.Creator<T>): T where T : Parcelable {
+        val dataParcel = Parcel.obtain()
+        dataParcel.unmarshall(byteArray, 0, byteArray.size)
+        dataParcel.setDataPosition(0)
+        val result = creator.createFromParcel(dataParcel)
+        dataParcel.recycle()
+        return result
+    }
+}
+
+suspend fun DownloadHelper.prepare() {
+    suspendCoroutine { cont ->
+        prepare(object : DownloadHelper.Callback {
+            override fun onPrepared(helper: DownloadHelper) {
+                cont.resumeWith(Result.success(Unit))
+            }
+
+            override fun onPrepareError(helper: DownloadHelper, e: IOException) {
+                cont.resumeWith(Result.failure(e))
+            }
+        })
+    }
+}
+
+class DownloadProgress(
+    private val downloads: MutableMap<String, DownloaderApi.Download> = mutableMapOf(),
+    private val progress: MutableMap<String, Double> = mutableMapOf()
+) {
+    fun add(download: DownloaderApi.Download, initialProgress: Double) {
+        downloads[download.key] = download
+        progress[download.key] = initialProgress
+    }
+
+    fun set(id: String, newProgress: Double): DownloadStatusChangedEvent? {
+        val download = downloads[id] ?: throw Exception("Unknown download key $id") // TODO: Better exception
+        val currentProgress = progress[id] ?: throw Exception("Unknown download key $id") // TODO: Better exception
+
+        if (newProgress >= 1.0) {
+            download.isFinished = true
+        }
+
+        if (currentProgress != newProgress) {
+            return null
+        }
+
+        progress[id] = newProgress
+
+        return DownloadStatusChangedEvent.Builder()
+            .setDownload(download)
+            .setProgress(newProgress)
+            .build()
+    }
+}
+
+class Downloader(private val context: Context) {
+    private val downloadManager: DownloadManager = createDownloadManager(context)
```

Comment:
dont we already create a downloadmanager in DownloadService.kt?
Isnt it sketchy to have two of this?

[Docs say this:](https://developer.android.com/reference/androidx/media3/exoplayer/offline/DownloadManager)
Normally a download manager should be accessed via a [DownloadService](https://developer.android.com/reference/androidx/media3/exoplayer/offline/DownloadService). When a download manager is used directly instead, downloads will be initially paused and so must be resumed by calling [resumeDownloads](https://developer.android.com/reference/androidx/media3/exoplayer/offline/DownloadManager#resumeDownloads()).

Reply:
You're right, we should probably create it once and assign it to a static var. Right now we do this for `SimpleCache`, but that can be removed if we also store the downloadmanager.

Maybe better idea to use an kotlin `object` instead of a static top-level var (that is how its done right now with `SimpleCache`)

About the docs: Not sure if there are ways to avoid calling `resumeDownloads` every time. That would be nice. But I did not find any other ways to query the downloadmanager.